Add support for fast pruning of inlined functions

fvannee opened the pull request:
Inlineable functions used to be slow to plan, because the query
preprocessing function could not find the relations inside the
functions, as they haven't been inlined yet at that point.
This commit adds a separate check in the get_relation_info_hook
to optimize pruning of hypertables.

For example, see current behavior on master for a hypertable with 1000 chunks:
Master:
- inlined function: ~22ms planning time
- same query but outside of function: ~0.25ms

With patch, the planning time is identical for both cases.

```
CREATE TABLE test (a int, b bigint NOT NULL);
SELECT create_hypertable('public.test', 'b', chunk_time_interval=>10);
INSERT INTO test SELECT i, i FROM generate_series(1, 10000) i;

CREATE OR REPLACE FUNCTION test_f(_ts bigint)
RETURNS SETOF test LANGUAGE SQL STABLE
as $f$
   SELECT DISTINCT ON (a) * FROM test WHERE b >= _ts AND b <= _ts + 2
$f$;

EXPLAIN(ANALYZE)
SELECT * FROM test_f(5);

Unique  (cost=8.60..8.65 rows=11 width=12) (actual time=0.035..0.041 rows=3 loops=1)
  ->  Sort  (cost=8.60..8.63 rows=11 width=12) (actual time=0.034..0.035 rows=3 loops=1)
        Sort Key: test.a
        Sort Method: quicksort  Memory: 25kB
        ->  Append  (cost=0.00..8.41 rows=11 width=12) (actual time=0.023..0.028 rows=3 loops=1)
              ->  Seq Scan on test  (cost=0.00..0.00 rows=1 width=12) (actual time=0.011..0.011 rows=0 loops=1)
                    Filter: ((b >= '5'::bigint) AND (b <= '7'::bigint))
              ->  Index Scan using _hyper_281_6387_chunk_test_b_idx on _hyper_281_6387_chunk  (cost=0.15..8.35 rows=10 width=12) (actual time=0.011..0.013 rows=3 loops=1)
                    Index Cond: ((b >= '5'::bigint) AND (b <= '7'::bigint))
Planning Time: 22.022 ms
Execution Time: 0.102 ms

EXPLAIN(ANALYZE)
SELECT DISTINCT ON (a) * FROM test WHERE b >= 5 AND b <= 5 + 2;

Unique  (cost=8.52..8.57 rows=10 width=12) (actual time=0.017..0.022 rows=3 loops=1)
  ->  Sort  (cost=8.52..8.54 rows=10 width=12) (actual time=0.016..0.018 rows=3 loops=1)
        Sort Key: _hyper_281_6387_chunk.a
        Sort Method: quicksort  Memory: 25kB
        ->  Index Scan using _hyper_281_6387_chunk_test_b_idx on _hyper_281_6387_chunk  (cost=0.15..8.35 rows=10 width=12) (actual time=0.007..0.009 rows=3 loops=1)
              Index Cond: ((b >= 5) AND (b <= 7))
Planning Time: 0.242 ms
Execution Time: 0.049 ms
```

k-rus commented:
@fvannee Thank you for contributing.
There is quite a big difference how optimiser works between PG 11 and PG 12. I see that the test in the CI failed on all version before PG12. I guess you made your fix on PG12.
Also can you run formatting? It is done my calling `make format`.

fvannee commented:
I'll investigate the issues with earlier PG versions and do the formatting.

fvannee commented:
It turns out not to make sense to add support for this pre-v12. The planner improvements in PG12 are what makes it possible to delay this optimization until `get_relation_info_hook`. In all other versions, the `simple_rte_rel_array` is already expanded by the time `get_relation_info_hook` is called, so we're too late for any optimizations.
I changed the PR to add an `#if PG12_GE` to reflect this.

fvannee commented:
Ok, tests pass now

fvannee commented:
Rebased on current master (with @svenklemm fix for #1875 )

fvannee commented:
Rebased again.
Thoughts on the PR?

fvannee commented:
Any feedback?

mkindahl commented on test/sql/plan_hypertable_inline.sql:
```diff
@@ -0,0 +1,24 @@
+-- plans must be the same in both cases
+-- specifically, the first plan should not contain the parent hypertable
+-- as that is a sign the pruning was not done successfully
+:PREFIX SELECT * FROM test_f(5);
+
+:PREFIX SELECT DISTINCT ON (a) * FROM test WHERE b >= 5 AND b <= 5 + 2;
```
There are a few negative tests missing here: cases where the pruning should *not* be done.

* This was discussed briefly above, but neither of these statements trigger a situation where `resultRelation == 0` and `commandType != CMD_SELECT` so if that is a situation that can occur, it would be good to have a test that you do not mark it for expansion.

* Since CTEs should not be expanded, please add that as a test case to make sure that it works.

* Similarly for queries with row marks (`SELECT ... FOR UPDATE` and friends). Please add tests to make sure that they are not expanded even if they should without the row marks.

fvannee replied:
Would you like to have these tests explicitly in this file? These code paths are actually triggered by existing tests in TimescaleDb (see my comment above for one example).

fvannee replied:
Perhaps to elaborate a bit more - the code path added allows to do pruning of inlineable functions, but the logic `is this a query where a function has been inlined` is not used anywhere in the code. Rather, by just postponing the marking of which tables need pruning to a later stage, when inlining has taken place already, we can just do our pruning as usual. This is why existing *negative* tests actually touch all of these code paths already. They haven't been selected for pruning during the first try in the planner preprocess hook, so they'll get another try in this hook. Whether or not there's an inlineable function doesn't change anything for the negative tests.

fvannee commented on src/planner.c:
```diff
@@ -796,6 +796,27 @@ timescaledb_get_relation_info_hook(PlannerInfo *root, Oid relation_objectid, boo
+			 * rte->requiredPerms if we're requesting an UPDATE/DELETE on this rel.
+			 */
+			if (!ts_guc_disable_optimizations && ts_guc_enable_constraint_exclusion && inhparent &&
+				rte->ctename == NULL && !IS_UPDL_CMD(query) && query->resultRelation == 0 &&
```
It happens in `planner.c` in Postgres source, around line 1300:

```
		/*
		 * Make a deep copy of the parsetree for this planning cycle to mess
		 * around with, and change it to look like a SELECT.  (Hack alert: the
		 * target RTE still has updatedCols set if this is an UPDATE, so that
		 * expand_partitioned_rtentry will correctly update
		 * subroot->partColsUpdated.)
		 */
		subroot->parse = copyObject(root->parse);

		subroot->parse->commandType = CMD_SELECT;
		subroot->parse->resultRelation = 0;
```

Indeed that makes `!IS_UPDL_CMD(query)` check redundant, because that'll always evaluate to true on v12. However, since I'd say Postgres applies a bit of a hack for planning update/delete's in v12, I think it's good to check anyway and not rely on the Postgres hack completely. If you disagree I'd be happy to remove it though.

Queries that trigger this behavior can actually be found in TimescaleDb tests already. For example, when running the tests without the extra ACL permission check, regressions come up. Here's one for an `UPDATE`, but a similar test exists for `DELETE`.

```
-- ConstraintAwareAppend NOT applied for UPDATE
EXPLAIN (costs off)
UPDATE "one_Partition"
SET series_1 = 8
WHERE series_1 IN (SELECT series_1 FROM "one_Partition" WHERE series_1 > series_val());
                                                                   QUERY PLAN                                                                   
------------------------------------------------------------------------------------------------------------------------------------------------
 Update on _hyper_1_1_chunk <<<<------- incorrect
   Update on _hyper_1_1_chunk
   Update on _hyper_1_2_chunk
   Update on _hyper_1_3_chunk
   ->  Hash Join
         Hash Cond: (_hyper_1_1_chunk.series_1 = "one_Partition".series_1)
         ->  Seq Scan on _hyper_1_1_chunk
         ->  Hash
               ->  HashAggregate
                     Group Key: "one_Partition".series_1
etc...
```

fvannee replied:
After some experimenting it would seem I cannot move them from the if-statement to the assert unfortunately.
The Postgres planner code I gave you is only part of the full, rather complicated, inheritance planner. The first step of the planner is what I showed - where it simulates it as a `SELECT` to find out which partitions to process. However, it then uses the results of this to modify its own `UPDATE`/`DELETE` plan and then runs it through the planner again. This time, it actually *resets* the `requiredPerms` field to `0` though.
This means that:
- In the first part of the planner, `commandType==SELECT` and `requiredPerms` are set to `UPDATE/DELETE`
- In the second part, it's called again but now with `commandType==UPDATE/DELETE` and `requiredPerms==0`.
The reset happens in `inherit.c` function `expand_single_inheritance_child`

These points together means we do need to check all conditions in the if-statement.
I will add some extra comments to this part though.

mkindahl replied:
Ok, then they need to be kept, but please add tests that make sure that nobody accidentally removes them and triggers a regression. That is, a test that will fail if any of the checks are removed, but not otherwise.

fvannee replied:
Yes, I'm adding extra tests now. Will post a new version soon.

fvannee commented on src/planner.c:
```diff
@@ -796,6 +796,27 @@ timescaledb_get_relation_info_hook(PlannerInfo *root, Oid relation_objectid, boo
+			 */
+			if (!ts_guc_disable_optimizations && ts_guc_enable_constraint_exclusion && inhparent &&
+				rte->ctename == NULL && !IS_UPDL_CMD(query) && query->resultRelation == 0 &&
+				query->rowMarks == NIL && (rte->requiredPerms & (ACL_UPDATE | ACL_DELETE)) == 0)
```
See my comment above

fvannee commented:
I've added some extra tests and rebased on latest master. Please let me know if the tests look alright now - especially the CTE tests I wasn't sure what exactly would be best to test here.

fvannee commented:
> @fvannee Thank you for contributing! It looks good! However, I have one request for change:
> 
> Since the output of `plan_hypertable_cache.sql` is different between PG versions, new SQL files are generated for each PG version from `plan_hypertable_cache.sql.in`. The generated files should be excluded from the PR and added to local `.gitignore` located in the same folder instead. Can you fix it?

Sure, what about the `scratch_file` and `scratch_file2` that are generated by CMake? Should they also be in a `.gitignore` somewhere? It's not part of this PR I guess but they show up after every CMake in the `sql` directory.

k-rus commented:
> Sure, what about the `scratch_file` and `scratch_file2` that are generated by CMake? Should they also be in a `.gitignore` somewhere? It's not part of this PR I guess but they show up after every CMake in the `sql` directory.

@fvannee I've never experienced them. I run `bootstrap` in the root of TimescaleDB and all generated files go into `build` folder, which is already ignored.

fvannee commented:
> > Sure, what about the `scratch_file` and `scratch_file2` that are generated by CMake? Should they also be in a `.gitignore` somewhere? It's not part of this PR I guess but they show up after every CMake in the `sql` directory.
> 
> @fvannee I've never experienced them. I run `bootstrap` in the root of TimescaleDB and all generated files go into `build` folder, which is already ignored.

For me, this is the output of a fresh clone followed by bootstrap:

```
florisvannee@localhost:~$ git clone https://github.com/timescale/timescaledb.git
Cloning into 'timescaledb'...
remote: Enumerating objects: 3, done.
remote: Counting objects: 100% (3/3), done.
remote: Compressing objects: 100% (3/3), done.
remote: Total 31642 (delta 0), reused 1 (delta 0), pack-reused 31639
Receiving objects: 100% (31642/31642), 11.93 MiB | 27.14 MiB/s, done.
Resolving deltas: 100% (26146/26146), done.
florisvannee@localhost:~$ cd timescaledb
florisvannee@localhost:~/timescaledb$ ./bootstrap
-- The C compiler identification is GNU 8.2.1
-- Check for working C compiler: /opt/rh/devtoolset-8/root/usr/bin/cc
-- Check for working C compiler: /opt/rh/devtoolset-8/root/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- TimescaleDB version 2.0.0-dev. Can be updated from version 1.7.1
-- Build type is Release
-- Install method is 'source'
-- Performing Test CC_SUPPORTS_IMPLICIT_FALLTHROUGH
-- Performing Test CC_SUPPORTS_IMPLICIT_FALLTHROUGH - Success
-- Performing Test CC_SUPPORTS_VISIBILITY_HIDDEN
-- Performing Test CC_SUPPORTS_VISIBILITY_HIDDEN - Success
-- Using compiler GNU
-- Found Git: /home/florisvannee/bin/git (found version "2.25.0") 
-- Using pg_config /usr/local/pgsql/12/bin/pg_config
-- Compiling against PostgreSQL version 12.0
-- Found PostgreSQL source in /home/florisvannee/workspace/postgres
CMake Warning at CMakeLists.txt:246 (message):
  clang-format version 7 or 8 required


CMake Warning at CMakeLists.txt:255 (message):
  clang-format is disabled (can't find clang-format or docker)


CMake Warning at CMakeLists.txt:281 (message):
  Did not find clang-tidy, not defining clang-tidy target.


-- Assertion checks are OFF
-- Found OpenSSL: /usr/lib64/libcrypto.so (found version "1.0.2k")  
-- OPENSSL_LIBRARIES: /usr/lib64/libssl.so;/usr/lib64/libcrypto.so
-- Using OpenSSL version 1.0.2k
-- Using nm /opt/rh/devtoolset-8/root/usr/bin/nm
-- Using pg_regress /usr/local/pgsql/12/lib/pgxs/src/test/regress/pg_regress
-- Using pg_isolation_regress /home/florisvannee/workspace/postgres/src/test/isolation/pg_isolation_regress
-- Configuring done
-- Generating done
-- Build files have been written to: /home/florisvannee/timescaledb/build
TimescaleDB build system initialized in ./build. To compile, do:
cd ./build && make
florisvannee@localhost:~/timescaledb$ git status
On branch master
Your branch is up to date with 'origin/master'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	sql/scratch_file
	sql/scratch_file2

nothing added to commit but untracked files present (use "git add" to track)
```

fvannee commented:
Regression tests seem to be quite flaky, however it's mostly for older versions that are completely untouched in this PR. Is this a known issue?

k-rus commented:
> Regression tests seem to be quite flaky, however it's mostly for older versions that are completely untouched in this PR. Is this a known issue?

@fvannee I restarted the failed build. I confirm that there are number of flaky tests and fixing them is in our backlog. Unfortunately, no low hanging fruits left there.

fvannee commented:
> > Regression tests seem to be quite flaky, however it's mostly for older versions that are completely untouched in this PR. Is this a known issue?
> 
> @fvannee I restarted the failed build. I confirm that there are number of flaky tests and fixing them is in our backlog. Unfortunately, no low hanging fruits left there.

Ok thanks, it's passing now :-)

fvannee commented:
Rebased again - I had to change the `pg_hba.conf.in` file that's used by regression tests to remove references to SSL, otherwise I couldn't run the tests locally (when compiling Postgres and TimescaleDb without SSL support). Was it intentional to break no-SSL support for the tests?

fvannee commented on tsl/test/expected/dist_hypertable-12.out:
```diff
@@ -2309,38 +2309,25 @@ UPDATE disttable_replicated SET device = 2 WHERE device = (SELECT device FROM de
-                             Output: _hyper_6_17_dist_chunk_1.device
-                             Data node: data_node_3
-                             Remote SQL: SELECT device FROM _timescaledb_internal._hyper_6_17_dist_chunk ORDER BY device ASC NULLS LAST
+                 ->  Custom Scan (AsyncAppend) (actual rows=1 loops=1)
```
Could someone explain to me why queries involving insert/update are not optimized in distributed hypertables? It seems to me that this plan looks fine too. This is one regression I saw when rebasing on distributed hypertables branch.
@mkindahl @k-rus

k-rus replied:
cc @erimatnor

fvannee replied:
Yes, I think your explanation is correct. Then I will keep the test case like this.
Are you OK with the current version of the code that still does the marking during preprocessing for PG12 as well, or do you want it to be changed in this PR such that PG12 *only* does the marking of hypertables in the get_relation_hook? To change this, there'd need to be additional #ifdef's in preprocess_query that exclude PG12 specifically.

erimatnor replied:
For the sake of being cautious, I think we can refactor expansion handling for PG12 later if that is really what we'd like to do. Feel free to take a look at it though.

fvannee replied:
Ok, will leave that for a future improvement then

fvannee commented on test/pg_hba.conf.in:
```diff
@@ -3,14 +3,14 @@
-hostssl  all             @TEST_ROLE_CLUSTER_SUPERUSER@  127.0.0.1/32           cert clientcert=1
-hostssl  all             @TEST_ROLE_1@                  127.0.0.1/32           cert clientcert=1
-hostssl  all             @TEST_ROLE_2@                  127.0.0.1/32           cert clientcert=1
+host     all             @TEST_ROLE_CLUSTER_SUPERUSER@  127.0.0.1/32           trust
```
> Rebased again - I had to change the pg_hba.conf.in file that's used by regression tests to remove references to SSL, otherwise I couldn't run the tests locally (when compiling Postgres and TimescaleDb without SSL support). Was it intentional to break no-SSL support for the tests?

> Similarly, I'm unable to run the TSL tests on Postgres builds without SSL support because of the postgresql.conf.in file in tsl/test directory that sets ssl=on. I haven't changed that in the PR though, as changing that also changed the port number that data nodes were running on, which in turn changes a lot of the expected test output.

I cannot run test cases locally on Postgresql builds that are built without SSL support. I would actually expect that when I build TimescaleDb without SSL, that I could then run the tests on a Postgres server without SSL support (why else would I build TimescaleDb without SSL support). I'm ok with changing it back, but I think the implications of this should be clear.

erimatnor replied:
We'd like to run tests with SSL since that will be our recommended setup for multinode configurations. The testing will be incomplete without it. Still, since this is a generated file, it should be possible to select auth method based on whether SSL is enabled in CMake or not and generate a working pg_hba.conf.

fvannee replied:
I'll revert the change here then

fvannee commented:
Rebased again. I think it's ready for merging. Could you let me know if anything else is expected to be changed before merge?

fvannee commented:
Thanks for approving. Rebased again - the failing test is the remote_txn test that was flaky according to #1961 , but I can't retrigger the run myself.

fvannee commented:
Pushed another rebase.. Changes on master are coming in faster than I can rebase. :-) Please advise when it can be merged.

svenklemm commented:
> Pushed another rebase.. Changes on master are coming in faster than I can rebase. :-) Please advise when it can be merged.

This needs another pass now since PG9.6 and PG10 support got removed in master plan_hypertable_cache probably doesnt need to be a template file anymore. Feel free to merge after cleaning up the test files.

fvannee commented:
Thanks @svenklemm I rebased on latest master and removed 9.6/10 test file. It cannot be merged though, because the output of 11 and 12 differs.
I cannot merge myself as I don't have write access, but anyone else feel free to merge when CI tests have finished.

svenklemm commented:
Thank you for your contribution.
